src/DotNetty.Codecs.Protobuf/ProtobufEncoder.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace DotNetty.Codecs.Protobuf
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.Contracts;
+    using DotNetty.Buffers;
+    using DotNetty.Transport.Channels;
+    using Google.Protobuf;
+
+    public sealed class ProtobufEncoder : MessageToMessageEncoder<IMessage>
+    {
+        protected override void Encode(IChannelHandlerContext context, IMessage message, List<object> output)
+        {
+            Contract.Requires(context != null);
+            Contract.Requires(message != null);
+            Contract.Requires(output != null);
+
+            IByteBuffer buffer = null;
+
+            try
+            {
+                int size = message.CalculateSize();
+                if (size == 0)
+                {
+                    return;
+                }
+
+                buffer = Unpooled.WrappedBuffer(message.ToByteArray());
```

Comment:
It turns out it is not much efficient even with some sort of wrapper output stream
This is the ToByteArray implementation

byte[] flatArray = new byte[message.CalculateSize()];
CodedOutputStream output = new CodedOutputStream(flatArray);
message.WriteTo(output);

This is the stream based implementation, internally still uses byte array
public CodedOutputStream(Stream output, int bufferSize, bool leaveOpen)
     : this(output, new byte[bufferSize], leaveOpen)
	 
The ToByteArray flat implementation actually is better because the stream based still requires another array